lib/Dialect/ESI/ESIServices.cpp

```diff
@@ -87,38 +208,45 @@ LogicalResult ESIConnectServicesPass::process(hw::HWMutableModuleLike mod) {
   for (auto instOp : modBlock.getOps<ServiceInstanceOp>())
     localImplReqs[instOp.service_symbolAttr()] = new Block();
 
-  // Classify each to_client req, appending non-local reqs to the list xor move
-  // into the block associated with the matching server.
-  SmallVector<RequestToClientConnection, 4> nonLocalToClientReqs;
+  // Find all of the "local" requests.
   mod.walk([&](RequestToClientConnection req) {
     auto service = req.servicePortAttr().getModuleRef();
     auto implOpF = localImplReqs.find(service);
-    if (implOpF == localImplReqs.end())
-      nonLocalToClientReqs.push_back(req);
-    else
+    if (implOpF != localImplReqs.end())
       req->moveBefore(implOpF->second, implOpF->second->end());
   });
-
-  // Classify each to_server req, appending non-local reqs to the list xor move
-  // into the block associated with the matching server.
-  SmallVector<RequestToServerConnection, 4> nonLocalToServerReqs;
   mod.walk([&](RequestToServerConnection req) {
     auto service = req.servicePortAttr().getModuleRef();
     auto implOpF = localImplReqs.find(service);
-    if (implOpF == localImplReqs.end())
-      nonLocalToServerReqs.push_back(req);
-    else
+    if (implOpF != localImplReqs.end())
       req->moveBefore(implOpF->second, implOpF->second->end());
   });
 
-  // Replace each service instance with a generation request.
+  // Replace each service instance with a generation request. If a service
+  // generator is registered, generate the server.
   for (auto instOp :
        llvm::make_early_inc_range(modBlock.getOps<ServiceInstanceOp>())) {
     Block *portReqs = localImplReqs[instOp.service_symbolAttr()];
     if (failed(replaceInst(instOp, portReqs)))
       return failure();
   }
 
+  // Identify the non-local reqs which need to be surfaced from this module.
+  SmallVector<RequestToClientConnection, 4> nonLocalToClientReqs;
+  SmallVector<RequestToServerConnection, 4> nonLocalToServerReqs;
+  mod.walk([&](RequestToClientConnection req) {
+    auto service = req.servicePortAttr().getModuleRef();
+    auto implOpF = localImplReqs.find(service);
+    if (implOpF == localImplReqs.end())
+      nonLocalToClientReqs.push_back(req);
+  });
+  mod.walk([&](RequestToServerConnection req) {
```

Comment:
Not sure how much this matters, but this is making two walks over the IR, and I don't think that's necessary here. Could you do a single walk with a TypeSwitch that looks for RequestToClientConnection or RequestToServerConnection?

Reply:
Yeah, I thought about that but ultimately decided that this was the more straightforward way to do it. If the block is long, though, we don't want to walk it twice so I'll recode it.